benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,944 @@
+import argparse
+import copy
+import json
+import os
+import statistics
+import sys
+import threading
+from pathlib import Path
+import pandas as pd
+import torch
+import torch.distributed as c10d
+import torch.distributed.rpc as rpc
+import torch.multiprocessing as mp
+import torch.nn as nn
+import torch.nn.functional as F
+from tabulate import tabulate
+from torch.distributed.rpc import TensorPipeRpcBackendOptions
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+
+GLOO = "gloo"
+NCCL = "nccl"
+INDICES = "indices"
+VALUES = "values"
+SIZE = "size"
+GRADIENT = "gradient"
+HOOK_METRIC = "hook_metric"
+FORWARD_METRIC = "foward_metric"
+BACKWARD_METRIC = "backward_metric"
+BATCH_LEVEL_METRIC = "batch_level_metric"
+GRADIENT_SERVER_BATCH_METRIC = "gradient_server_batch_metric"
+GRADIENT_SERVER_STRAGGLER_METRIC = "gradient_server_straggler_metric"
+RANK = "rank"
+TRAINER_COUNT = "trainer_count"
+USE_CUDA_RPC = "use_cuda_rpc"
+BATCH_MODE = "batch_mode"
+CUDA_SPARSE_RPC = "cuda_sparse_rpc"
+CPU_SPARSE_RPC = "cpu_sparse_rpc"
+CUDA_DENSE_RPC = "cuda_dense_rpc"
+CPU_DENSE_RPC = "cpu_dense_rpc"
+NCCL_ALLREDUCE_DENSE = "nccl_allreduce_dense"
+NCCL_ALLREDUCE = "nccl_allreduce"
+GLOO_ALLREDUCE = "gloo_allreduce"
+BATCH_ALL = "batch_all"
+FORWARD_PASS = "forward_pass"
+BACKWARD = "backward"
+BP_LOC_STRAGGLER = "bp_loc_straggler"
+BP_LOC_BATCH = "bp_loc_batch"
+
+
+def get_name(rank, ddp_trainers, gradient_servers):
+    if rank < ddp_trainers:
+        return "trainer{}".format(rank)
+    elif rank < (ddp_trainers + gradient_servers):
+        return "gs{}".format(rank)
+    else:
+        return "master"
+
+
+def gs_for_rank(rank, configurations):
+    if (configurations.ddp_trainers % configurations.gradient_servers) != 0:
+        gs_rank = rank % configurations.gradient_servers
+    else:
+        x = int(configurations.ddp_trainers / configurations.gradient_servers)
+        gs_rank = int(rank / x)
+    return gs_rank + configurations.ddp_trainers
+
+
+def record_event(rank, metric_type, key, name, metrics):
```

Comment:
nit: this function modifies states that belongs to another class. It might make more sense to have a dedicated `Metrics` class that owns the metrics data.